OCPSTRAT-2485: Introduce feature gate based inclusion/exclusion of manifests

JoelSpeed opened the pull request:
This PR introduces a new annotation `release.openshift.io/feature-gate` that accepts a comma separated list of feature gates which must be enabled to consider a manifest for inclusion within the cluster, optionally preceeded by a `-` to negate that (ie don't include me if this gate is enabled).

The PR adds a feature gate informer to the core operator which keeps a synchronised list of enabled feature gates. This is then passed through to sync workers, added to the sync worker status, and finally passed into the library-go manifest inclusion/exclusion code for decisions on whether the manifest should/shouldn't be included.

I need to test this and explore what the render and task graph parts are about. I see some todo's in the render already about proper feature-setting so it looks like maybe we are ok to ignore that for now.

Note, claude was used to assist in the development of this, I have reviewed every line carefully so I _think_ this is all good and follows existing CVO patterns, but I'm not a CVO expert so please point out where Claude and I have misunderstood 🙏 

Draft for now as this is based on a library-go PR that I need to raise and merge before this can go in.

JoelSpeed commented:
/test e2e-aws-ovn-techpreview

As a smoke test for generally not breaking existing stuff

JoelSpeed commented:
Have been testing this on a cluster and locally applying the rendering and I'm confident this is now doing as it's supposed to. I think this is ready for review @DavidHurta 

I have yet to look at how this affects HyperShift, that's next

DavidHurta commented on hack/cluster-version-util/task_graph.go:
```diff
@@ -30,7 +31,7 @@ func newTaskGraphCmd() *cobra.Command {
 func runTaskGraphCmd(cmd *cobra.Command, args []string) error {
 	manifestDir := args[0]
-	release, err := payload.LoadUpdate(manifestDir, "", "", "", payload.DefaultClusterProfile, nil)
+	release, err := payload.LoadUpdate(manifestDir, "", "", "", payload.DefaultClusterProfile, nil, sets.Set[string]{})
```
I do not think so. It would be nice, but as you have stated, the file is not being kept up-to-date with new functionalities, so I believe it's okay not to implement it for now.

DavidHurta commented on pkg/cvo/featuregate_integration_test.go:
```diff
@@ -0,0 +1,280 @@
+			name:         "include manifest with matching feature gate",
+			enabledGates: sets.New[string]("TechPreviewFeatureGate"),
+			manifestAnnotations: map[string]string{
+				"release.openshift.io/feature-gates": "TechPreviewFeatureGate",
```
My take, for what it's worth. Anything that felt more natural when creating tests or when testing manifests :bow: The developers will be the users, and both options can be understood. I also like the idea of removing the s. However, that's just my personal opinion after reviewing the code.

JoelSpeed replied:
I've adjusted the library-go PR to account for the dropped S, and will update this one (and clean up the commit history what a mess Joel) once that has merged

DavidHurta commented on pkg/cvo/cvo.go:
```diff
@@ -1084,6 +1100,97 @@ func (optr *Operator) HTTPClient() (*http.Client, error) {
+}
+
+// updateEnabledFeatureGates updates the cluster feature gates based on a FeatureGate object
+func (optr *Operator) updateEnabledFeatureGates(obj interface{}) bool {
```
return bool description

DavidHurta commented on pkg/cvo/cvo.go:
```diff
@@ -1084,6 +1100,97 @@ func (optr *Operator) HTTPClient() (*http.Client, error) {
+	}
+
+	// If no matching version found, log a warning but continue with empty set
+	if enabledGates.Len() == 0 {
```
"log warning" but InfoF

DavidHurta commented on pkg/cvo/cvo.go:
```diff
@@ -189,6 +191,10 @@ type Operator struct {
 	// featurechangestopper controller will detect when cluster feature gate config changes and shutdown the CVO.
 	enabledFeatureGates featuregates.CvoGateChecker
```
also the description and de facto definition needs clarification

DavidHurta commented on pkg/cvo/cvo.go:
```diff
@@ -1084,6 +1100,97 @@ func (optr *Operator) HTTPClient() (*http.Client, error) {
+}
+
+// initializeFeatureGates initializes the cluster feature gates from the current FeatureGate object
+func (optr *Operator) initializeFeatureGates() {
```
if featuregates can be configured during install time; the - semantic could be configured to not allow some manifests to be deployed;l however, the try logic would deploy them if defaulted to "no enabled gates" before being synced; this could potentially introduce some issue in the future?? maybe, the hard failure could become more determistic

DavidHurta commented on pkg/cvo/cvo.go:
```diff
@@ -189,6 +191,10 @@ type Operator struct {
 
+	// featureGatesMutex protects access to enabledManifestFeatureGates
+	featureGatesMutex           sync.RWMutex
+	enabledManifestFeatureGates sets.Set[string]
```
The naming of the `enabledFeatureGates` and `enabledManifestFeatureGates` fields is now not distinct. WDYT? Maybe we could rename `enabledFeatureGates` to something like `enabledCVOFeatureGates` to make their nature more explicit. Otherwise, it's not that easy to know the difference between them at first glance.

DavidHurta commented on pkg/cvo/cvo.go:
```diff
@@ -318,7 +334,7 @@ func (optr *Operator) LoadInitialPayload(ctx context.Context, restConfig *rest.C
 	}
 
 	update, err := payload.LoadUpdate(optr.defaultPayloadDir(), optr.release.Image, optr.exclude, string(optr.requiredFeatureSet),
```
One thing that has come to my mind. With the proposed changes, the CVO could be now working with two different FeatureGate generations at a moment. The feature set is received a quite differently than the feature gates. Given the `Include()` logic allows for funky logic like:

```
annotations:
   features-set: TechPreviewNoUpgrade
   feature-gate: BlowClusterUpWhenDefaultSetIsEnabled
```

to state the desire to apply a manifest only when a desired feature set is set AND when a desired feature gate is enabled/disabled.

DavidHurta replied:
I may also be misunderstanding the usage of the annotations 👀

DavidHurta commented on pkg/cvo/cvo.go:
```diff
@@ -1084,6 +1100,97 @@ func (optr *Operator) HTTPClient() (*http.Client, error) {
+	// If no matching version found, log a warning but continue with empty set
+	if enabledGates.Len() == 0 {
+		klog.V(2).Infof("No feature gates found for current version %s, using empty set", currentVersion)
+	}
```
The comment states `log a warning`, but the utilized `klog` uses `Infof` and not `Warningf`

DavidHurta reviewed (commented):
Hey, Joel 👋 🙇The PR looks solid, I appreciate you taking the time to implement it. A few comments. I am mainly a bit concerned regarding the potential non-deterministic application of manifests, which was not apparent to me initially (see that wall of text comment). Apologies for taking a little bit longer. The PR is quite large, and some things were not clear to me from the commit history. Feel free to disagree with me and/or verify my statements (the CVO can sometimes be more complex than it maybe should be).

JoelSpeed commented on bootstrap/bootstrap-pod.yaml:
```diff
@@ -5,6 +5,8 @@ metadata:
   labels:
     k8s-app: cluster-version-operator
+  annotations:
+    include.release.openshift.io/{{ .ClusterProfile }}: "true"
```
Yeah I can update that. This was a result of me actually implementing the manifest filtering completely through the render step. Previously it didn't actually do most filtering which meant that manifests without a cluster profile annotation were applied, now, we need a cluster profile annotation on all files to be rendered, and this one was missing it

JoelSpeed commented on pkg/cvo/cvo.go:
```diff
@@ -1084,6 +1100,97 @@ func (optr *Operator) HTTPClient() (*http.Client, error) {
+
+// initializeFeatureGates initializes the cluster feature gates from the current FeatureGate object
+func (optr *Operator) initializeFeatureGates() {
+	// Try to load initial state from the cluster FeatureGate object
```
So we already have this covered in `pkg/start/start.go` which fetches the initial gates and feature set. If the object doesn't exist, we assume nothing, and I'm adding a comment about why that is ok for feature gates. If there's a genuine error it returns and the startup will fail and CVO will exit by the looks of it.

So by the time we get here, we either didn't see the FeatureGate because it didn't exist, or there's a genuine error, in which case, given this is backed by an informer, would have come up in start.go and we would have exited there.

I think this is ok as is given the earlier protection.

JoelSpeed commented:
Thanks for the review again @DavidHurta, I've rebased and squashed a few of the smaller commits down. The last 5 commits are based on your most recent round of feedback, but I'll squash those into some of the larger commits logically once you are happy with those

@JianLi-RH should be able to test now, no conflicts at the moment

DavidHurta commented:
Hi, @JoelSpeed :wave: Everything looks great. Feel free to address the failing `ci/prow/verify-yaml` job, squash the commits as you have mentioned, and we are good to go :bow:

JianLi-RH commented:
Still could not create cluster by clusterbot:
The command I used to start the cluster is:
```
launch 4.22,openshift/cluster-version-operator#1273 aws
```
Log: https://storage.googleapis.com/test-platform-results/logs/release-openshift-origin-installer-launch-aws-modern/2018521245519187968/build-log.txt

```
Building github.com/openshift/cluster-version-operator cluster-version-operator binary (v1.0.0-1582-g3d8c3e5c-dirty)
# github.com/openshift/cluster-version-operator/pkg/cvo
pkg/cvo/cvo.go:1185:14: optr.enabledFeatureGates undefined (type *Operator has no field or method enabledFeatureGates)
cp: cannot stat '_output/linux/amd64/cluster-version-operator': No such file or directory
[36mINFO[0m[2026-02-03T03:26:10Z] Ran for 9m38s                                
[31mERRO[0m[2026-02-03T03:26:10Z] Some steps failed:                           
[31mERRO[0m[2026-02-03T03:26:10Z] 
  * could not run steps: step cluster-version-operator failed: error occurred handling build cluster-version-operator-amd64: the build cluster-version-operator-amd64 failed after 6m54s with reason DockerBuildFailed: Dockerfile build strategy has failed. 
 ```

JianLi-RH commented:
/verified by @JianLi-RH
